consensus/core/src/linearizer.rs

```diff
@@ -0,0 +1,489 @@
+                let commit_block = commit_block_opt
+                    .expect("We should have the block referenced in the commit data");
+                if commit_block.reference() == commit_data.leader {
+                    leader_block_idx = Some(idx);
+                }
+                commit_block
+            })
+            .collect::<Vec<_>>();
+        let leader_block_idx = leader_block_idx.expect("Leader block must be in the sub-dag");
+        let leader_block_ref = blocks[leader_block_idx].reference();
+        let timestamp_ms = blocks[leader_block_idx].timestamp_ms();
+        CommittedSubDag::new(leader_block_ref, blocks, timestamp_ms, commit_data.index)
+    }
+
+    /// Sort the blocks of the sub-dag by round number. Any deterministic algorithm works.
+    pub fn sort(&mut self) {
+        self.blocks.sort_by_key(|x| x.round());
+    }
+}
+
+impl Display for CommittedSubDag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "CommittedSubDag(leader={}, index={}, blocks=[",
+            self.leader, self.commit_index
+        )?;
+        for (idx, block) in self.blocks.iter().enumerate() {
+            if idx > 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", block.digest())?;
+        }
+        write!(f, "])")
+    }
+}
+
+impl fmt::Debug for CommittedSubDag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}@{} (", self.leader, self.commit_index)?;
+        for block in &self.blocks {
+            write!(f, "{}, ", block.reference())?;
+        }
+        write!(f, ")")
+    }
+}
+
+/// Expand a committed sequence of leader into a sequence of sub-dags.
+#[allow(unused)]
+pub struct Linearizer {
+    /// In memory block store representing the dag state
+    dag_state: Arc<RwLock<DagState>>,
+}
+
+#[allow(unused)]
+impl Linearizer {
+    pub fn new(dag_state: Arc<RwLock<DagState>>) -> Self {
+        Self { dag_state }
+    }
+
+    /// Collect the sub-dag from a specific leader excluding any duplicates or
+    /// blocks that have already been committed (within previous sub-dags).
+    fn collect_sub_dag(
+        &mut self,
+        leader_block: VerifiedBlock,
+        committed: &HashSet<BlockRef>,
+        last_commit_index: u64,
+        last_committed_rounds: Vec<u32>,
+    ) -> CommittedSubDag {
+        let mut to_commit = Vec::new();
+        let mut committed = committed.clone();
+
+        let timestamp_ms = leader_block.timestamp_ms();
+        let leader_block_ref = leader_block.reference();
+        let mut buffer = vec![leader_block];
+        assert!(committed.insert(leader_block_ref));
+        while let Some(x) = buffer.pop() {
+            to_commit.push(x.clone());
+
+            let ancestors: Vec<VerifiedBlock> = self
+                .dag_state
+                .read()
+                .get_uncommitted_blocks(
+                    x.ancestors()
```

Comment:
who are these ancestors? the parents? Are we sure the list is always complete (and that there can't be concurrent updates to the dag_state to add new ancestors)?